datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1470,34 +1471,67 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
             }) => Transformed::yes(simplify_regex_expr(left, op, right)?),
 
             // Rules for Like
-            Expr::Like(Like {
-                expr,
-                pattern,
-                negated,
-                escape_char: _,
-                case_insensitive: _,
-            }) if matches!(
-                pattern.as_ref(),
-                Expr::Literal(ScalarValue::Utf8(Some(pattern_str))) if pattern_str == "%"
-            ) || matches!(
-                pattern.as_ref(),
-                Expr::Literal(ScalarValue::LargeUtf8(Some(pattern_str))) if pattern_str == "%"
-            ) || matches!(
-                pattern.as_ref(),
-                Expr::Literal(ScalarValue::Utf8View(Some(pattern_str))) if pattern_str == "%"
-            ) =>
-            {
-                // exp LIKE '%' is
-                //   - when exp is not NULL, it's true
-                //   - when exp is NULL, it's NULL
-                // exp NOT LIKE '%' is
-                //   - when exp is not NULL, it's false
-                //   - when exp is NULL, it's NULL
-                Transformed::yes(Expr::Case(Case {
-                    expr: Some(Box::new(Expr::IsNotNull(expr))),
-                    when_then_expr: vec![(Box::new(lit(true)), Box::new(lit(!negated)))],
-                    else_expr: None,
-                }))
+            Expr::Like(like) => {
+                match as_string_scalar(&like.pattern) {
+                    Some((data_type, pattern_str)) => {
+                        match pattern_str {
+                            None => return Ok(Transformed::yes(lit_bool_null())),
+                            Some(pattern_str) if pattern_str == "%" => {
+                                // exp LIKE '%' is
+                                //   - when exp is not NULL, it's true
+                                //   - when exp is NULL, it's NULL
+                                // exp NOT LIKE '%' is
+                                //   - when exp is not NULL, it's false
+                                //   - when exp is NULL, it's NULL
+                                let result_for_non_null = lit(!like.negated);
+                                Transformed::yes(if !info.nullable(&like.expr)? {
+                                    result_for_non_null
+                                } else {
+                                    Expr::Case(Case {
+                                        expr: Some(Box::new(Expr::IsNotNull(like.expr))),
+                                        when_then_expr: vec![(
+                                            Box::new(lit(true)),
+                                            Box::new(result_for_non_null),
+                                        )],
+                                        else_expr: None,
+                                    })
+                                })
+                            }
+                            Some(pattern_str)
```

Comment:
I am not sure about this rule (see questions below) -- maybe you could add some comments explanation on what it is supposed to do.

Reply:
```suggestion
                            Some(pattern_str)
                                // Repeated occurrences of wildcard are redundant so remove them
                                // exp LIKE '%%'  --> exp LIKE '%'
```